Followup to TCP changes in pr/40817

gmarcosb opened the pull request:
#### Summary

This is a followup to https://github.com/project-chip/connectedhomeip/pull/40817 with additional clean-up, formalizing ref-counting & removing visiblity into internals, so that users are pushed to use auto ref counting

#### Testing

CI

gmarcosb committed:
Fix ActiveTCPConnectionState being leaked when CASE connections are closed (e.g. via ExpireAllSessionsOnLogicalFabric)

gmarcosb committed:
Support multiple connections to same peer by sharing a single TCP socket

gmarcosb committed:
Unit test with multiple connections, to simulate SessionManager

gmarcosb committed:
Use automatic ref-counting throughout; limit visibility of explicit ref/deref

gmarcosb commented on src/lib/support/CodeUtils.h:
```diff
@@ -245,12 +246,12 @@
  *  @param[in]  expr        A scalar expression to be evaluated against CHIP_NO_ERROR.
  */
-#define LogAndReturnErrorOnFailure(expr, MOD, MSG, ...)                                                                            \
+#define ReturnErrorAndLogOnFailure(expr, MOD, MSG, ...)                                                                            \
```
Continuing discussion @bzbarsky-apple : I just noticed that `ReturnLogErrorOnFailure`, which you suggested we could combine, also logs `__FILE__` and `__LINE__` explicitly; do we really want that? i.e. the ChipError will already have that in debug builds, seems a bit much to also log file/line where we caught the error...

If you agree that's unnecessary, I can convert `ReturnLogErrorOnFailure` to use this

gmarcosb committed:
Address review comments from https://github.com/project-chip/connectedhomeip/pull/40817/files

gmarcosb committed:
Address review comments from https://github.com/project-chip/connectedhomeip/pull/40817/files

gmarcosb committed:
Additional review comments & fixes

gmarcosb commented on src/inet/TCPEndPointImplSockets.cpp:
```diff
@@ -734,6 +736,10 @@ CHIP_ERROR TCPEndPointImplSockets::GetSocket(IPAddressType addrType)
         {
             return CHIP_ERROR_POSIX(errno);
         }
+        auto connectionCleanup = ScopeExit([&]() {
```
@bzbarsky-apple pretty sure this was the cause of the `SEGV on unknown address ` you saw; basically if the port is already in use, `StartWatchingSocket` below fails & it all falls apart; it would be nice if our test could test for this generically by simply reserving the port before starting the test, which theoretically would cause this failure; I'm unsure how to do that

gmarcosb committed:
Check for invalid nullptr input

gmarcosb committed:
* Tighten/hide usage of ref-counting internals, using new method names to catch wrong usages of manual ref counting
* Use a consistent Free() for all inet connections which does clean-up such as closing the connections before freeing memory back to pool

gmarcosb committed:
Apply clang tidy suggestions

gmarcosb committed:
Apply review suggestions

gmarcosb committed:
Undo simplification; this was likely intended to force  memory allocation on the matter thread

gmarcosb committed:
Apply review suggestions

gmarcosb committed:
Go back to using && and std::move as previous impl did (EndPointHandle now has a move operator/ctor)

bzbarsky-apple reviewed (approved):
Approved, assuming https://github.com/project-chip/connectedhomeip/pull/40904#discussion_r2421620271 happens in a followup soon.

gmarcosb committed:
Make call to HandleAcceptError explicit, as that's all we're doing anyway, and set mAppState
Remove unnecessary initialization as it's now done @ top of method
Fixes CI failure:

==68889==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000030 (pc 0x000104884514 bp 0x00016b642490 sp 0x00016b642410 T0)
==68889==The signal is caused by a READ memory access.
==68889==Hint: address points to the zero page.
    #0 0x104884514 in chip::Transport::TCPBase::FindActiveConnection(chip::Inet::EndPointHandle<chip::Inet::TCPEndPoint> const&) TCP.cpp:205
    #1 0x1048826a8 in chip::Transport::TCPBase::HandleAcceptError(chip::Inet::EndPointHandle<chip::Inet::TCPEndPoint> const&, chip::ChipError) TCP.cpp:664
    #2 0x10488181c in chip::Transport::TCPBase::HandleIncomingConnection(chip::Inet::EndPointHandle<chip::Inet::TCPEndPoint> const&, chip::Inet::EndPointHandle<chip::Inet::TCPEndPoint> const&, chip::Inet::IPAddress const&, unsigned short) TCP.cpp:623
    #3 0x1048b3328 in chip::Inet::TCPEndPointImplSockets::HandleIncomingConnection() TCPEndPointImplSockets.cpp:1066
    #4 0x1048b13ac in chip::Inet::TCPEndPointImplSockets::HandlePendingIO(chip::BitFlags<chip::System::SocketEventFlags, unsigned char>) TCPEndPointImplSockets.cpp:797
    #5 0x1048a96a8 in chip::Inet::TCPEndPointImplSockets::HandlePendingIO(chip::BitFlags<chip::System::SocketEventFlags, unsigned char>, long) TCPEndPointImplSockets.cpp:783
    #6 0x10493f870 in chip::System::LayerImplDispatch::SocketWatch::HandleEvents(chip::System::LayerImplDispatch::SelectSets const&) const SystemLayerImplDispatchSockets.mm:245
    #7 0x10493e800 in chip::System::LayerImplDispatch::SocketWatchPool::HandleEvents(chip::System::LayerImplDispatch::SelectSets const&) const SystemLayerImplDispatchSockets.mm:204
    #8 0x10493df84 in chip::System::LayerImplDispatch::HandleSocketsAndTimerEvents(std::__1::chrono::duration<unsigned int, std::__1::ratio<1l, 1000l>>) SystemLayerImplDispatchSockets.mm:172
    #9 0x104930e28 in chip::System::LayerImplDispatch::HandleDispatchQueueEvents(std::__1::chrono::duration<unsigned int, std::__1::ratio<1l, 1000l>>) SystemLayerImplDispatch.mm:298

gmarcosb committed:
Fix errors merging to head

gmarcosb committed:
Apply code review suggestion
